Add support for grants on new entity kinds

djsstarburst opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

This PR privides a first-cut at syntax generalizing
privilege grants so that the entity kinds to which
the grants apply and the privileges for those grants
can be extended.
    
The PR also adds APIs to grant, deny, revoke and list grants
on arbitrary entity kinds.  Entity kinds TABLE and SCHEMA
are processed as before, but new entity kinds can be
supported by implementing the new methods on SystemAccessControl
and SystemSecurityMetadata.


<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

martint commented on core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4:
```diff
@@ -838,7 +838,17 @@ pathSpecification
+
+entityKind
+    : TABLE | SCHEMA
+    | LETTER+ (UNDERSCORE LETTER+)*
```
This uses lexer tokens that are defined as "fragments" and are not visible to the parser. I suspect that as it currently exists, this production never matches. Do you have any tests that involve an entity kind that's not table or schema?

djsstarburst replied:
I fixed the fragment creating the appropriate productions.

HOWEVER, when I built a test in `TestSqlParser` for exotic privileges and entity kinds, I immediately ran into an ambiguity between `GRANT role TO grantee` and `GRANT funky_privilege ON t TO grantee`.  Before my changes, the ambiguity doesn't  exist, because the privileges, `SELECT`, `CREATE`, etc, are keywords and therefore don't collide with role names.

I don't know how to fix this ambiguity.  Do you have any suggestions?

djsstarburst replied:
I refactored the grammar and all tests now pass.

martint commented on core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:
```diff
@@ -0,0 +1,91 @@
+import static java.util.Objects.requireNonNull;
+
+public class GrantScope
+        extends Statement
```
This should be a Node, not a Statement

djsstarburst replied:
Changed.

martint commented on core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:
```diff
@@ -0,0 +1,91 @@
+    private final Optional<String> entityKind;
+    private final QualifiedName name;
+
+    public GrantScope(Optional<String> entityKind, QualifiedName name)
```
This constructor should not be needed. There's no scenario where the object should be instantiated without a location.

djsstarburst replied:
Removed.

martint commented on core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:
```diff
@@ -0,0 +1,91 @@
+        this(Optional.of(location), entityKind, name);
+    }
+
+    public GrantScope(Optional<NodeLocation> location, Optional<String> entityKind, QualifiedName name)
```
Same here.

djsstarburst replied:
Removed.

martint commented on core/trino-parser/src/main/java/io/trino/sql/tree/GrantScope.java:
```diff
@@ -0,0 +1,91 @@
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class GrantScope
```
I'd call this `GrantObject`, as it's meant to identify the object for which privileges are being granted. FWIW, the SQL spec calls this construct an "object name", so the name I'm suggesting matches that more closely.

djsstarburst replied:
Changed.

djsstarburst commented:
I believe this is in shape for review.  I refactored the grammar and all tests now pass.

martint commented on core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:
```diff
@@ -119,35 +119,44 @@ statement
-        TO grantee=principal
-        (WITH GRANT OPTION)?                                           #grant
+    | GRANT (
+        (
```
Have you tried using two separate labels for each of the branches instead of bundling them under `#genericGrant` ?

djsstarburst replied:
I did try.  They problem I saw an error in the IDE when was #grant inside the required outer parens.  AFAICT #tags are only allowed at the top level.

martint replied:
Then, I would try the following:

```
| GRANT privilegeOrRole (',' privilegeOrRole)*
            TO principal (',' principal)*
            (WITH ADMIN OPTION)?
            (GRANTED BY grantor)?
            (IN catalog=identifier)?   #grantRole
| GRANT ((privilegeOrRole (',' privilegeOrRole)*) | ALL PRIVILEGES)
            ON grantObject
            TO principal
            (WITH GRANT OPTION)? #grantPrivilege
```

djsstarburst replied:
Awesome, @martint, that worked and all parser tests pass!

I pushed the improved version.

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java:
```diff
@@ -0,0 +1,45 @@
+public class EntityKindAndName
+{
+    private final String entityKind;
+    // TODO: Should this instead be List<String> like QualifiedName?
```
Yes, but it depends on what we're trying to achieve. Flattening a multipart QualifiedName into a string is problematic due to lack of escaping of special characters. For instance a name such as "a.b".c.d cannot be flattened to a string "a.b.c.d" without loss of information. If we want to be able to recover the catalog, schema and name, we need to keep them separate.

In fact. We may just want to separate catalog/schema/name as different elements.

djsstarburst replied:
Let me know what you prefer, @martint, and I'll change it to conform.

djsstarburst replied:
I decided that `EntityKindAndName` should use `List<String>` as the name, and made that change.

I also added `SystemSecurityMetadata.getAllEntityKindPrivileges` to provide a way to know at parse time if a privilege is legal.

Finally, I created [this galaxy-trino branch](https://github.com/starburstdata/galaxy-trino/tree/djs/exotic-privileges) to support the new Trino entity privilege APIs once they are merged.

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:
```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
+    {
+        List<String> privileges;
+        if (context.ALL() != null) {
+            privileges = ImmutableList.of();
```
Why empty list? Just use `null` for this to remove any ambiguity about "empty means all" vs "empty means none"

djsstarburst replied:
Changed.

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:
```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
     @Override
     public Node visitGrantRoles(SqlBaseParser.GrantRolesContext context)
     {
+        // Check legality for a role grant.
```
Is this a statement about what this code is doing? If so, I don't see it. Or is it a "TO DO"?

djsstarburst replied:
It was a leftover from the generic case.  Removed.

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:
```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
+    @Override
+    public Node visitRevokePrivileges(SqlBaseParser.RevokePrivilegesContext context)
+    {
+        List<String> privileges = extractPrivileges(context.privilegeOrRole());
```
How is this method handling "ALL PRIVILEGES" ? Is empty list a side effect of the rule matching "ALL PRIVILEGES" instead of privilegeOrRole? If so, let's make the check explicit.

Also, see my comment above re: the semantics of empty list.

djsstarburst replied:
It was missing the required test for "ALL PRIVILEGES".  Added, and handled the empty list as suggested.

djsstarburst commented on core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:
```diff
@@ -916,6 +916,16 @@ sqlStatementList
 
 privilege
     : CREATE | SELECT | DELETE | INSERT | UPDATE
+    | IDENTIFIER
```
Privilege formatting changed.  I picked `IDENTIFIER` because I thought we would be unlikely to need or want quoted or digit identifiers.  But we likely want to allow `nonReserverd` tokens to be privileges, so `identifier` is the better choice.  Changed.

djsstarburst commented on core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4:
```diff
@@ -916,6 +916,16 @@ sqlStatementList
+
+entityKind
+    : TABLE | SCHEMA
+    | IDENTIFIER
```
Both changed.

djsstarburst commented on core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java:
```diff
@@ -0,0 +1,45 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class EntityKindAndName
```
LOL.  When I originally wrote the PR, Trino didn't yet allow records.

Changed to a record.

djsstarburst commented on core/trino-spi/src/main/java/io/trino/spi/connector/EntityPrivilege.java:
```diff
@@ -0,0 +1,36 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class EntityPrivilege
```
Changed to a record.

djsstarburst commented on core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java:
```diff
@@ -3382,85 +3376,123 @@ public void testDeny()
+
+    private void testExoticPrivilegeGrants(String privilege, Optional<String> entityKind, PrincipalSpecification.Type type)
+    {
+        assertStatement(formatGrant("GRANT %s ON %st TO %su", privilege, entityKind, type),
```
Unfortunately, `assertThat(statement(...))` requires accurate determination of the location of element, but these are tests whose SQL strings are constructed, so it is nontrivial to come up with the exact locations.

I just checked, and there are 352 uses of `assertStatement` and 81 uses of `assertThat(statement(...))`.  Unless you can think of a good way to come up with the locations, it seems better to me to allow `assertStatement` here.

martint replied:
We're trying to get rid of `assertStatement`. If we keep adding test cases that use it, we just make the migration harder for whoever ends up doing it. Why not do it the right way in the first place?

> requires accurate determination of the location of element, but these are tests whose SQL strings are constructed, so it is nontrivial to come up with the exact locations.

Then we should structure the tests in a different way so that we don't run into that problem. Let me take a closer look at what they are doing.

djsstarburst replied:
I removed the generated tests, and wrote out the important tests long-hand using `assertThat(statement(...)).isEqualTo(...)`.

djsstarburst commented:
The support for entity kind privileges was not handling ALL PRIVILEGES for exotic entity kinds.  I fixed that by adding a method on `Metadata` and `SystemSecurityMetadata` to get all privileges for an exotic entity kind.  That method is below.  I was conflicted about whether the new method should be added to `SystemAccessControl` or `SystemSecurityMetadata`.  In the end, I added it to `SystemSecurityMetadata`:

```
    /**
     * Gets all the EntityPrivileges associated with an entityKind.  Defines ALL PRIVILEGES
     * for the entityKind
     */
    default Set<EntityPrivilege> getAllEntityKindPrivileges(String entityKind)
    {
        throw new UnsupportedOperationException();
    }
```

mosabua commented:
I assume you are still working on this @djsstarburst ...

martint commented on core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java:
```diff
@@ -44,6 +48,24 @@ public static Set<Privilege> parseStatementPrivileges(Node statement, Optional<L
+                    .map(privilege -> {
+                        EntityPrivilege entityPrivilege = new EntityPrivilege(privilege.toUpperCase(Locale.ENGLISH));
+                        if (!allPrivileges.contains(entityPrivilege)) {
+                            throw new ParsingException("Privilege %s is not supported for entity kind %s".formatted(privilege, entityKind));
```
This should not be a ParsingException, but a TrinoException with INVALID_PRIVILEGE error code.

djsstarburst replied:
Changed.

martint commented on core/trino-main/src/main/java/io/trino/execution/PrivilegeUtilities.java:
```diff
@@ -44,6 +48,24 @@ public static Set<Privilege> parseStatementPrivileges(Node statement, Optional<L
         return privileges;
     }
 
+    public static Set<EntityPrivilege> parseEntityKindPrivileges(String entityKind, Metadata metadata, Optional<List<String>> privileges)
```
This method is misnamed. There's no parsing involved here. I would rename it to "fetchEntityPrivileges" or similar.

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:
```diff
@@ -1640,72 +1674,20 @@ public Node visitDeny(SqlBaseParser.DenyContext context)
                 getLocation(context),
-                context.TABLE() != null,
-                tableName);
+                context.grantObject() == null || context.grantObject().isEmpty() ? Optional.empty()
+                        : Optional.of(createGrantObject(getLocation(context), context.grantObject())));
```
Format as follows for readability (i.e., break the line at the `||` instead of at the `:` to make the visual layout more closely align with the evaluation precedence)

```suggestion
                context.grantObject() == null || 
                context.grantObject().isEmpty() ? Optional.empty() : Optional.of(createGrantObject(getLocation(context), context.grantObject())));
```

djsstarburst replied:
Changed.

martint commented on core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java:
```diff
@@ -3113,7 +3095,7 @@ public Node visitGroupingOperation(SqlBaseParser.GroupingOperationContext contex
 
     @Override
-    public Node visitUnquotedIdentifier(SqlBaseParser.UnquotedIdentifierContext context)
+    public Node visitUnquotedIdentifier(UnquotedIdentifierContext context)
```
Unrelated change?

djsstarburst replied:
Reverted.

martint commented on core/trino-parser/src/main/java/io/trino/sql/tree/GrantObject.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(getEntityKind(), getName());
```
Use the fields directly instead of going through the public getters, here and in equals(). While not the case today, it's conceivable that the getters might return a derived view of the internal state of the object that doesn't match the internal state.

djsstarburst replied:
Good point, and a lesson I'll remember in the future.  Changed.

martint commented on core/trino-spi/src/main/java/io/trino/spi/connector/EntityKindAndName.java:
```diff
@@ -0,0 +1,29 @@
+    public EntityKindAndName
+    {
+        requireNonNull(entityKind, "entityKind is null");
+        name = List.copyOf(requireNonNull(name, "name is null"));
```
requireNonNull not needed for this one. List.copyOf will fail if `name` is null.

djsstarburst replied:
Removed.

martint reviewed (commented):
Just a few minor comments, but otherwise looks good.

djsstarburst commented:
Thanks very much for reviewing, @martint!  I made the changes you suggested.

mosabua commented:
💥  @djsstarburst !

djsstarburst commented:
Yay, it's merged!
